README.md
# ARM Debugger
An easy to use, powerful debugger.  You can import a rom, and an sgm file, and set breakpoints, watchpoints, and readpoints, view/modify registers and memory, output results to a txt file, disassemble the code, and more!


## Set Up
Make sure you have Python installed, and download the repository.  You can then unpack it and open "Debugger.py", and that's it!
From there, you can type in `help` or `?` to get detailed info on the available commands.
You can also start Debugger.py from the command line with two optional arguments: `[filepath to rom] [filepath to savestate]`

Don't have Python? Download the exe **and** settings file here:  
https://drive.google.com/open?id=12rcQcuH55hC64-DNFB7QGVVT5y6R6HHs

To execute ROM instructions, you'll need to import a ROM.  You can use the command `importrom [filepath]`.  You can upload savestate files that are in .sgm format using `importstate [filepath]`.
```
> importrom C:\Users\GenericName\Games\Roms\romname.gba
ROM loaded successfully
> importstate C:\Users\GenericName\Games\Saves\statename.sgm
State loaded successfully

# works with or without quotation marks, regardless of spaces in the path name, 
# so clicking and dragging a file to the command window will work after typing importrom/importstate
```
Settings may be changed in the `Debugger_Settings.txt` file.  You can change the default output files, output formats, the file size limit, or the initial registers.  You can also write debugger commands to execute whenever the program starts up.

If you ever get stuck in an infinite loop, press `ctrl + c` to escape it.

That's pretty much all you need to get started.  The rest of this readme is just here to explain some features.


## Basic Commands
- `n (count)` - execute *count* instruction(s), displaying the registers.  Count=1 by default.
- `c (count)` - execute *count* instruction(s). Count=infinity by default.
- `b [addr]` - set a breakpoint at *addr*.  CPU execution will halt after *addr* is executed.
    - if *addr* is "all", displays all break/write/readpoints
- `bw [addr]` - set a watchpoint.  CPU execution will halt after *addr* has been written to.
- `br [addr]` - set a readpoint.  CPU execution will halt after *addr* has been read from.
- `bc [condition]` - set a conditional breakpoint.  CPU execution will halt if *condition* is true.
- `d [addr]` - delete a breakpoint
    - if *addr* is "all", deletes all break/write/readpoints
- `dw [addr]` - delete a watchpoint
- `dr [addr]` - delete a readpoint
- `dc [index]` - delete a conditional breakpoint by index number
- `i` - print the CPU registers
- `dist [addr] (count)` - display *count* THUMB instructions starting from *addr*
- `disa [addr] (count)` - display *count* ARM instructions starting from *addr*
- `m [addr] (bytecount) (size)` - display *bytecount* bytes of memory at *addr* (count=1, size=1 by default)
- `asm (-varname) (addr): (command)` - assemble a command written in Thumb
    - if *varname* is included, the commands will be stored in the User Variable *varname* as a byte string
    - if *addr* is included, you can utilize absolute address references, like `bl $08014878` or `ldr r0, [$08014894]`
    - if *command* is omitted, it enters multiline mode, allowing you to paste multiple commands
- `disasm [code]` - disassembles a single 16-bit number into a Thumb instruction
    - if *code* is a byte string, this command can disassemble multiple instructions
- `fbounds [addr]` - detects and displays the boundaries of the Thumb function containing *addr*

**Enter in nothing to execute the previous command.**  

Arguments are separated by spaces.  Arguments in brackets `[...]` are required, and parentheses `(...)` are optional.  
All commands accept [*expressions*](#expressions) as arguments, however if the command accepts multiple arguments, the expressions must not contain spaces.


## Expressions
With this debugger, you can utilize and modify registers and memory directly as though they were python variables.
You can assign values to variables by typing in: *name* = *expression*
```
test = 12345678   # creates a user variable named 'test'
```
Assigning a value to registers and memory is just as easy
```
r0 = test   # sets the actual register r0 equal to test
m($02000000) = r0   # sets the memory at $02000000 equal to r0
m($08000000) = bytestring  # using a bytestring, like b'example', you can easily overwite multiple bytes of memory
```
Default variables are **r0-r16, sp, lr, pc,** and **m(*addr*, *size*)**.  (size=4 by default).  r16 is CPSR  
There are also 4 global variables that may be accessed, but not directly modified:
- `MODE` - 0 if in ARM mode, 1 if in THUMB mode 
- `ADDR` - The current address
- `INSTR` - The next machine code instruction to be executed
- `CPUCOUNT` - The total number of CPU instructions executed since the beginning of the session

Attempts to modify these (or any other) global variables will instead create a User Variable with the same name.  
In [Execution Mode](#execution-mode), these and any other global variables may be modified.


You can also modify variables with compound assignment operators. (+=, -=, \*=, etc).  `m(r1) += r2`  
Expressions may include any combination of variables and mathematical operations.  
Expressions may be typed directly into to the console to print their value.  
Hexadecimal numbers must be preceded by "0x" or "$". 
```
base = $08000000
m(base + 0xc, 8) = 4*r0 + r1*r2 - m(r3)//4
```
**You can use expressions in place of any arguments**

If the command takes multiple arguments, then each expression must not contain spaces.
```
> chardata = $02000520
> m chardata+$14C*4 10
02000A50:  696C6546 00000078 00000000 05000000   Felix...........
02000A60:  00220046 3ACF4000 000C0020 0102001A   F."..@.: .......
02000A70:  00000000 0077006D                     ....m.w.
```


## Higher Level Commands
- `if [condition]: [command]` - execute *command* if *condition* is true
- `while [condition]: [command]` - repeat *command* while *condition* is true
- `rep/repeat [count]: [command]` - repeat *command* *count* times
- `[name][op][expression]` - create/modify a User Variable
- `def [name]: [commands]`
    - bind a list of commands separated by semicolons to *name*
    - execute these functions later by typing in "*name*()"
    - you can call functions within functions, with unlimited nesting
- `tree [addr] (depth)` - prints a tree of functions based on what functions are called in Thumb mode
- `save (name)` - create a local save; *name* = PRIORSTATE by default
- `load (name)` - load a local save; *name* = PRIORSTATE by default
- `dv [name]` - delete user variable
- `df [name]` - delete user function
- `ds (name)` - delete local save; *name* = PRIORSTATE by default
- `vars` - print all user variables
- `funcs` - print all user functions
- `saves` - print all local saves  

**You can write multiple commands in a single line by separating them with `;`**  
**You can use multiple-command if/while/repeat instructions by separating each inner command with `..`**  
(using a semicolon will end the if/while/repeat instruction)  
The commands may be anything, including function calls, loops, and even [Debugger Mode Swaps](#alternate-debugger-modes).
```
> wram = $02000000; m wram 32
02000000:  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00   ................
02000010:  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00   ................
> iter = 0; rep 32: m(wram + iter) = iter .. iter += 1; m wram 32
02000000:  00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F   ................
02000010:  10 11 12 13 14 15 16 17 18 19 1A 1B 1C 1D 1E 1F   ................
> def clear: iter = 0; while iter < arg1: m(arg0 + iter, 1) = 0 .. iter += 1
> arg0 = wram; arg1 = 16; clear()
> m wram 32
02000000:  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00   ................
02000010:  10 11 12 13 14 15 16 17 18 19 1A 1B 1C 1D 1E 1F   ................
> funcs
{'clear': iter = 0; while iter < arg1: m(arg0 + iter, 1) = 0 .. iter += 1}
```  

### Local Saves
`save` and `load` only apply to local saves.  Local saves are temporarily stored in the current session.  They can be given names and loaded with those names at any time.  To overwrite an actual savestate file, use `exportstate`.  


## File and OS Commands
- `importrom [filepath]` - import a rom into the debugger
- `importstate [filepath]` - import a savestate
- `exportstate (filepath)` - save the current state to a file; 
    - *filepath* = (most recent import) by default
    - will overwrite the destination, back up your saves!
- `exportrom (filepath)` - save the current ROM to a file; 
    - *filepath* = (most recent import) by default
    - will overwrite the destination, back up your saves!
- `reset` - reset the emulator (clears the RAM and resets the registers)
- `output [condition]`
    - after each CPU instruction, if *condition* is True, the debugger will write data to "Debugger_Output.txt"

Changelog.md
### November 17th, 2020
- added search function: Searches BIOS, RAM, and ROM
  - accepts integers, strings, and byte-like objects
  - for integers, can optionally specify a byte count
- spaces within quotation marks are no longer treated as separators
- fixed instructions that modify pc directly
- improved fbounds command, and added fboundsa for arm function boundaries
- added "nn" command, which is like "n", but doesn't step into functions
- modified "c" command; now you can specify an address to stop at
- regular breakpoints now halt prior to execution


### April 13th, 2020

- Fixed Overflow Flag calculation bug
- Modified error handling of Assembler
- Made assign not break with "=" sign in string
- Removed replacement of x with 0x
- Added ability to read/write BIOS
- Swapped replacement order in expstr_compile
- Added -out parameter to asm command, to store bytes of instructions
- Reworked strings to accept single quotes and escape sequences
- Added multi-byte disasm commands
- Made m command's 2nd argument be bytecount instead of element count
- Changed default arguments of m command
- Fixed "FileNotFound" error for default outputs
- Added option to display function with fbounds command


### April 6th, 2020

- pre-compiled expstr and assign functions
- fixed bitmask error with bl and blh instructions
- added ability to store strings, can now use byte strings
- changed concatenation operator to ".."
- modified exec mode to make error messages nicer
- added comtype3 for functions that don't split args
- fixed assignment to m(...) types
- added &=, |=, ^=, %= to MatchAssign
- can now use iterators instead of reversed lists in commandque
- made assembler match letters instead of non-spaces
- improved code in com_asm
- separated terminal command from output command
- assembly commands now accept // and () line comments
- fixed assembler bug mistaking "add rn, 13/15" for "add rn, sp/pc"
- added assembler error message
- added fbounds command
- added disasm command
- capitalized addresses in disassembler
- added custom print and input functions for terminal binding
- can now specify default folders
- improved file system error messages
- added dir, getcwd, and chdir os commands
- Completely rewrote assembler, eliminating reliance on eval();
  - now has more than twice as many lines of code, but is 3 times faster
- made disassembler ignore 0xb600-0xbbff range


Components/ARMCPU.py
from bisect import bisect_right


BIOS, RAM, ROM = bytearray(), bytearray(), bytearray()
REG = [0]*17
RegionMarkers = {}
BreakState = ""
BreakPoints = set()
WatchPoints = set()
ReadPoints = set()
Conditionals = []
Executing = False


def undef(*args): pass


def mem_read(addr,size=4,signed=False):
    region = addr >> 24 & 0xF
    if region in RegionMarkers:
        base, length = RegionMarkers[region]
        reladdr = (addr & 0xFFFFFF) % length + base
        value = int.from_bytes(RAM[reladdr:reladdr + size],"little")
    else:
        if region >= 8:
            reladdr = addr - 0x08000000
            value = int.from_bytes(ROM[reladdr:reladdr+size],"little")
        elif region == 0:
            reladdr = addr % 0x4000
            value = int.from_bytes(BIOS[reladdr:reladdr+size],"little")
        else: return 0
    if signed:
        msb = 2**(8*size-1)
        value = ((value^msb) - msb) & 0xFFFFFFFF
    if addr in ReadPoints and Executing:
        global BreakState
        BreakState = f"ReadPoint: ${addr:0>{2*size}X} (={value:0>{2*size}X})"
    return value


def mem_write(addr,data,size=4):
    region = addr >> 24 & 0xF
    if type(data) is int: data = int.to_bytes(data % 2**(8*size),size,"little")
    else: size = len(data)
    if addr in WatchPoints and Executing:
        global BreakState
        old, new = mem_read(addr,size), int.from_bytes(data, "little")
        BreakState = f"WatchPoint: {addr:0>8X} ({old:0>{2*size}X} -> {new:0>{2*size}X})"
    if region in RegionMarkers:
        base, length = RegionMarkers[region]
        reladdr = (addr & 0xFFFFFF) % length + base
        RAM[reladdr:reladdr+size] = data
        if RAM[RegionMarkers[4][0] + 0xDF] & 2**7: DMA()
    else:
        if region >= 8:
            reladdr = addr - 0x08000000
            ROM[reladdr:reladdr+size] = data
        elif region == 0:
            reladdr = addr % 0x4000
            BIOS[reladdr:reladdr+size] = data


def mem_copy(src,des,size):
    region1 = src >> 24 & 0xF
    region2 = des >> 24 & 0xF
    if region1 >= 8:
        reladdr = src - 0x08000000
        copydata = ROM[reladdr:reladdr + size]
    else:
        base, length = RegionMarkers[region1]
        reladdr = (src & 0xFFFFFF) % length + base
        copydata = RAM[reladdr:reladdr + size]
    try:
        base, length = RegionMarkers[region2]
        reladdr = (des & 0xFFFFFF) % length + base
        RAM[reladdr:reladdr+size] = copydata
    except KeyError:
        reladdr = des - 0x08000000
        ROM[reladdr:reladdr+size] = copydata


def DMA():
    src = mem_read(0x040000D4,4)
    des = mem_read(0x040000D8,4)
    count = mem_read(0x040000DC,2)
    control = mem_read(0x040000DE,2)
    mem_copy(src, des, count*(2 + 2*(control >> 10 & 1)))
    mem_write(0x040000DE, control & 0x7FFF, 2)


def compare(Op1,Op2,S=1):
    Op1 &= 0xFFFFFFFF
    Op2 &= 0xFFFFFFFF
    result = Op1 + Op2
    N = 8 if result & 2**31 else 0
    Z = 4 if not result & 0xFFFFFFFF else 0
    C = 2 if result & 2**32 else 0
    V = 1 if Op1 >> 31 == Op2 >> 31 != bool(N) else 0
    if S: REG[16] = (N|Z|C|V) << 28 | REG[16] & 2**28-1
    return result & 0xFFFFFFFF


def cmphalf(result,S=1):
    result &= 0xFFFFFFFF
    N = 8 if result & 2**31 else 0
    Z = 4 if not result else 0
    if S: REG[16] = (N|Z) << 28 | REG[16] & 2**30-1
    return result


def barrelshift(value,Shift,Typ,S=0,skipzero=False):
    value &= 0xFFFFFFFF
    Shift &= 31
    affectedflags = 0xE << 28
    C = 0
    if Shift: 
        C = 2*((value >> Shift-1) & 1)  # The last bit shifted out for rshifts
        if Typ == 0: value <<= Shift; C = 2*(value >> 32 & 1)
        elif Typ == 1: value >>= Shift
        elif Typ == 2: value = (value ^ 2**31) - 2**31 >> Shift
        elif Typ == 3: value = (value << 32 | value) >> Shift
    elif not skipzero:
        if Typ == 0: affectedflags = 0xC << 28
        elif Typ == 1: C = 2*(value>>31); value = 0
        elif Typ == 2: C = 2*(value>>31); value = -(value>>31)
        elif Typ == 3: affectedflags = 0xC << 28; value = ((REG[16] & 1<<29) << 3 | value) >> 1
    value &= 0xFFFFFFFF
    N = 8 if value & 2**31 else 0
    Z = 4 if not value else 0
    if S: REG[16] = REG[16] & ~affectedflags | ((N|Z|C) << 28) & affectedflags
    return value


conditions = (
    lambda cpsr: cpsr & 4 == 4,             # EQ: Z=1
    lambda cpsr: cpsr & 4 == 0,             # NE: Z=0
    lambda cpsr: cpsr & 2 == 2,             # CS/HS: C=1
    lambda cpsr: cpsr & 2 == 0,             # CC/LO: C=0
    lambda cpsr: cpsr & 8 == 8,             # MI: N=1
    lambda cpsr: cpsr & 8 == 0,             # PL: N=0
    lambda cpsr: cpsr & 1 == 1,             # VS: V=1
    lambda cpsr: cpsr & 1 == 0,             # VC: V=0
    lambda cpsr: cpsr & 6 == 2,             # HI: C=1 and Z=0
    lambda cpsr: cpsr & 6 != 2,             # LS: C=0 or Z=1
    lambda cpsr: cpsr & 9 in {0,9},         # GE: N=V
    lambda cpsr: cpsr & 9 not in {0,9},     # LT: N!=V
    lambda cpsr: cpsr & 13 in {0,9},        # GT: Z=0 and N=V
    lambda cpsr: cpsr & 13 not in {0,9},    # LE: Z=1 or N!=V
    lambda cpsr: True,                      # AL: Always true
    lambda cpsr: False,                     # NV: Never true
)


#######################
### THUMB FUNCTIONS ###
#######################


def shifted(instr):
    Op,Offset,Rs,Rd = instr>>11 & 3, instr>>6 & 31, instr>>3 & 7, instr & 7
    REG[Rd] = barrelshift(REG[Rs],Offset,Op,1)


def addsub(instr):
    I,Op,Rn,Rs,Rd = instr>>10 & 1, instr>>9 & 1, instr>>6 & 7, instr>>3 & 7, instr & 7
    Rs = REG[Rs]
    if not I: Rn = REG[Rn]
    if not Op: REG[Rd] = compare(Rs,Rn)
    else: REG[Rd] = compare(Rs,-Rn)
    

def immediate(instr):
    Op,Rd,Offset = instr>>11 & 3, instr>>8 & 7, instr & 0xFF
    Op1 = REG[Rd]
    if Op == 0: REG[Rd] = cmphalf(Offset)
    elif Op == 1: compare(REG[Rd], -Offset)
    elif Op == 2: REG[Rd] = compare(REG[Rd], Offset)
    elif Op == 3: REG[Rd] = compare(REG[Rd],-Offset)


alu_ops = (
    lambda Rd,Rs: cmphalf(Rd & Rs),                        # AND
    lambda Rd,Rs: cmphalf(Rd ^ Rs),                        # XOR
    lambda Rd,Rs: barrelshift(Rd,Rs & 0x1F,0,1,True),      # LSL
    lambda Rd,Rs: barrelshift(Rd,Rs & 0x1F,1,1,True),      # LSR
    lambda Rd,Rs: barrelshift(Rd,Rs & 0x1F,2,1,True),      # ASR
    lambda Rd,Rs: compare(Rd, Rs + (REG[16]>>29 & 1)),     # ADC
    lambda Rd,Rs: compare(Rd,-Rs + (REG[16]>>29 & 1) - 1), # SBC
    lambda Rd,Rs: barrelshift(Rd,Rs & 0x1F,3,1),           # ROR
    lambda Rd,Rs: cmphalf(Rd & Rs),                        # TST
    lambda Rd,Rs: compare(0,-Rs),                          # NEG
    lambda Rd,Rs: compare(Rd,-Rs),                         # CMP
    lambda Rd,Rs: compare(Rd, Rs),                         # CMN
    lambda Rd,Rs: cmphalf(Rd | Rs),                        # ORR
    lambda Rd,Rs: cmphalf(Rd * Rs),                        # MUL
    lambda Rd,Rs: cmphalf(Rd & ~Rs),                       # BIC
    lambda Rd,Rs: cmphalf(~Rs),                            # MVN
)

def AluOp(instr):
    Op,Rs,Rd = instr>>6 & 15, instr>>3 & 7, instr & 7
    result = alu_ops[Op](REG[Rd],REG[Rs])
    if Op not in {8,10,11}: REG[Rd] = result


def HiRegBx(instr):
    Op,Hd,Hs,Rs,Rd = instr>>8 & 3, instr>>7 & 1, instr>>6 & 1, instr>>3 & 7, instr & 7
    Rd += 8*Hd
    Rs += 8*Hs
    if Op == 0: REG[Rd] = (REG[Rd] + REG[Rs]) & 0xFFFFFFFF
    elif Op == 1: compare(REG[Rd],-REG[Rs])
    elif Op == 2: REG[Rd] = REG[Rs]
    elif Op == 3:
        Mode = REG[Rs] & 1
        if Hd: REG[14] = REG[15] + 1
        REG[15] = REG[Rs] + 4-3*Mode
        REG[16] = REG[16] & ~(1<<5) | Mode << 5
    if Rd == 15: REG[15] += 2


def ldr_pc(instr):
    Rd,Word = instr>>8 & 7, instr & 0xFF
    REG[Rd] = mem_read(REG[15] + Word*4 - (REG[15] & 2), 4)


def ldrstr(instr):
    Op,S,Ro,Rb,Rd = instr>>10 & 3, instr>>9 & 1, instr>>6 & 7, instr>>3 & 7, instr & 7
    addr = REG[Rb] + REG[Ro]
    if not S:
        if Op == 0: mem_write(addr, REG[Rd], 4)
        elif Op == 1: mem_write(addr, REG[Rd], 1)
        elif Op == 2: REG[Rd] = mem_read(addr, 4)
        elif Op == 3: REG[Rd] = mem_read(addr, 1)
    else:
        if Op == 0: mem_write(addr, REG[Rd], 2)
        elif Op == 1: REG[Rd] = mem_read(addr, 1, True)
        elif Op == 2: REG[Rd] = mem_read(addr, 2)
        elif Op == 3: REG[Rd] = mem_read(addr, 2, True)


def ldrstr_imm(instr):
    Op,L,Offset,Rb,Rd = instr>>12 & 3, instr>>11 & 1, instr>>6 & 31, instr>>3 & 7, instr & 7
    size = Op^2 if Op != 2 else 4  # (3,2,0) -> (1,4,2)
    addr = REG[Rb] + Offset*size
    if not L: mem_write(addr, REG[Rd], size)
    else: REG[Rd] = mem_read(addr, size)


def ldrstr_sp(instr):
    Op,Rd,Word = instr>>11 & 1, instr>>8 & 7, instr & 0xFF
    if not Op: mem_write(REG[13] + Word*4, REG[Rd], 4)
    else: REG[Rd] = mem_read(REG[13] + Word*4, 4)


def get_reladdr(instr):
    Op,Rd,Word = instr>>11 & 1, instr>>8 & 7, instr & 0xFF
    if not Op: REG[Rd] = (REG[15] & ~2) + Word*4
    else: REG[Rd] = REG[13] + Word*4
    REG[Rd] &= 0xFFFFFFFF


def add_sp(instr):
    S,Word = instr>>7 & 1, instr & 0x7F
    if not S: REG[13] += Word*4
    else: REG[13] -= Word*4
    REG[13] &= 0xFFFFFFFF


def pushpop(instr):
    Op,Rlist = instr>>11 & 1, instr & 0x1FF
    Rlist = [i for i in range(9) if Rlist & 2**i]
    if Rlist[-1] == 8: Rlist[-1] = 14 + Op
    if not Op:
        for i in reversed(Rlist):
            REG[13] -= 4
            mem_write(REG[13], REG[i], 4)
    else:
        for i in Rlist:
            REG[i] = mem_read(REG[13], 4)
            REG[13] += 4
        if Rlist[-1] == 15: REG[15] = (REG[15] & ~1) + 2
    REG[13] &= 0xFFFFFFFF


def stmldm(instr):
    Op,Rb,Rlist = instr>>11 & 1, instr>>8 & 7, instr & 0xFF
    addr = REG[Rb]
    for i in range(8):
        if Rlist & 2**i:
            if not Op: mem_write(addr, REG[i], 4)
            else: REG[i] = mem_read(addr, 4)
            addr += 4
    REG[Rb] = addr & 0xFFFFFFFF


def b_if(instr):
    Cond,Offset = instr>>8 & 15, instr & 0xFF
    if conditions[Cond](REG[16]>>28): 
        REG[15] = (REG[15] + ((Offset^0x80) - 0x80)*2 + 2) & 0xFFFFFFFF


def branch(instr):
    REG[15] = (REG[15] + ((instr & 0x7FF ^ 0x400) - 0x400)*2 + 2) & 0xFFFFFFFF


def bl(instr):
    if instr == 0xF800:
        link = REG[15] - 1
        REG[15] = REG[14] + 2
        REG[14] = link
    else:
        REG[14] = REG[15] + 1
        REG[15] = (REG[15] + (((instr & 0x7FF ^ 0x400) << 11 | (instr >> 16) & 0x7FF) - 0x200000)*2 + 2) & 0xFFFFFFFF


ThumbBounds = (
    0x1800,0x2000,0x4000,0x4400,0x4800,0x5000,0x6000,0x8000,0x9000,0xA000,
    0xB000,0xB400,0xBE00,0xC000,0xD000,0xDE00,0xDF00,0xE000,0xE800,0xF000,
)

ThumbFuncs = (
    shifted, addsub, immediate, AluOp, HiRegBx, ldr_pc, ldrstr, ldrstr_imm, ldrstr_imm, ldrstr_sp,
    get_reladdr, add_sp, pushpop, undef, stmldm, b_if, undef, undef, branch, undef, bl
)



#####################
### ARM FUNCTIONS ###
#####################


dataprocess_list = (
    lambda Rn,Op2,S: cmphalf(Rn & Op2, S),                         # AND
    lambda Rn,Op2,S: cmphalf(Rn ^ Op2, S),                         # EOR
    lambda Rn,Op2,S: compare(Rn, -Op2, S),                         # SUB
    lambda Rn,Op2,S: compare(Op2, -Rn, S),                         # RSB
    lambda Rn,Op2,S: compare(Rn, Op2, S),                          # ADD
    lambda Rn,Op2,S: compare(Rn, Op2 + (REG[16]>>29 & 1), S),      # ADC
    lambda Rn,Op2,S: compare(Rn,-Op2 + (REG[16]>>29 & 1) - 1, S),  # SBC
    lambda Rn,Op2,S: compare(Op2,-Rn + (REG[16]>>29 & 1) - 1, S),  # RSC
    lambda Rn,Op2,S: cmphalf(Rn & Op2, S),                         # TST
    lambda Rn,Op2,S: cmphalf(Rn ^ Op2, S),                         # TEQ
    lambda Rn,Op2,S: compare(Rn, -Op2, S),                         # CMP
    lambda Rn,Op2,S: compare(Rn, Op2, S),                          # CMN
    lambda Rn,Op2,S: cmphalf(Rn | Op2, S),                         # ORR
    lambda Rn,Op2,S: cmphalf(Op2, S),                              # MOV
    lambda Rn,Op2,S: cmphalf(Rn & ~Op2, S),                        # BIC
    lambda Rn,Op2,S: cmphalf(~Op2, S),                             # MVN
)


def dataprocess(instr):
    i = instr
    Imm, OpCode, S, Rn, Rd, Shift, Typ, Imm2, Rm = (
        i>>25 & 1, i>>21 & 15, i>>20 & 1, i>>16 & 15, i>>12 & 15, i>>7 & 31, i>>5 & 3, i>>4 & 1, i & 15)
    if Imm:
        Op2 = i & 0xFF
        Shift = (i >> 8 & 15)*2
        Op2 = ((Op2 & 2**Shift-1) << 32 | Op2) >> Shift
    else:
        if Imm2: Shift = REG[Shift>>1 & 15]
        Op2 = barrelshift(REG[Rm],Shift,Typ,S)
    result = dataprocess_list[OpCode](REG[Rn],Op2,S)
    if not(8 <= OpCode <= 11):
        REG[Rd] = result
        if Rd == 15: REG[15] += 4


def psr(instr):
    i = instr
    I, P, L, Field, Rd, Shift, Imm, Rm = (
        i>>25 & 1, i>>22 & 1, i>>21 & 1, i>>16 & 15, i>>12 & 15, i>>8 & 15, i & 0xFF, i & 15)
    if not P:
        if L:
            bitmask = 15<<28*(Field>>3) | 0xEF*(Field & 1)
            if I: Op = barrelshift(Imm, Shift*2, 3)
            else: Op = REG[Rm]
            REG[16] = REG[16] & ~bitmask | Op & bitmask
        else: R[Rd] = R[16]


def arm_bx(instr):
    L, Rn = instr >> 5 & 1, instr & 15
    if L: REG[14] = REG[15] - 4
    Mode = REG[Rn] & 1
    REG[15] = REG[Rn] + 4-3*Mode
    REG[16] = REG[16] & ~(1<<5) | Mode << 5


def arm_branch(instr):
    L, Offset = instr >> 24 & 1, (instr & 0xFFFFFF ^ 2**23) - 2**23
    if L: REG[14] = REG[15] - 4
    REG[15] = (REG[15] + 4 + Offset*4) & 0XFFFFFFFF


def clz(instr):
    Rd,Rm = instr >> 12 & 15, instr & 15    
    REG[Rd] = 32 - int.bit_length(REG[Rm])


def multiply(instr):
    i = instr
    OpCode, S, Rd, Rn, Rs, y, x, Rm = (i>>21 & 15, i>>20 & 1, i>>16 & 15, i>>12 & 15, i>>8 & 15, i>>6 & 1, i>>5 & 1, i & 15)
    Rm, Rs = REG[Rm], REG[Rs]
    if OpCode & 8:  #half registers
        OpCode &= 3
        S = 0
        Rs = (Rs >> 16*y & 0xFFFF ^ 2**15) - 2**15
        if OpCode != 1: Rm = (Rm >> 16*x & 0xFFFF ^ 2**15) - 2**15
        if OpCode == 0: REG[Rd] = (Rm*Rs + REG[Rn]) & 0xFFFFFFFF
        elif OpCode == 1: 
            Rm = (Rm ^ 2**31) - 2**31
            REG[Rd] = ((Rm*Rs >> 16) + REG[Rn]*(1-x)) & 0xFFFFFFFF
        elif OpCode == 2:
            result = (REG[Rd] << 32 | REG[Rn]) + Rm*Rs
            REG[Rd] = result >> 32 & 0xFFFFFFFF
            REG[Rn] = result & 0xFFFFFFFF
        elif OpCode == 3:
            REG[Rd] = Rm*Rs & 0xFFFFFFFF
    else:
        if OpCode & 2:  #signed
            Rm = (Rm ^ 2**31) - 2**31
            Rs = (Rs ^ 2**31) - 2**31
        result = Rm*Rs + REG[Rn]*(OpCode & 1)
        if OpCode & 4: 
            REG[Rd] = result >> 32 & 0xFFFFFFFF
            REG[Rn] = result & 0xFFFFFFFF
        else: 
            REG[Rd] = result & 0xFFFFFFFF
        if S:
            N = 8 if result & 2**63 else 0
            Z = 4 if not result & 2**63-1 else 0
            REG[16] = (N|Z)<<28 | (REG[16] & 2**30-1)


def datatransfer(instr):
    Flags, Rn, Rd, Offset = (instr>>20 & 0x7F, instr>>16 & 15, instr>>12 & 15, instr & 0xFFF)
    #Flags: Single/Double, Immediate offset, Pre/Post index, Up/Down, Byte/Int, Writeback, Load/Store
    D,I,P,U,B,W,L = (1 if Flags & 2**(6-i) else 0 for i in range(7))
    Shift, Typ, Rm = Offset>>7 & 31, Offset>>5 & 3, Offset & 15
    U = (-1,1)[U]
    addr = REG[Rn]
    if D:
        if I: Offset = barrelshift(REG[Rm],Shift,Typ)
        if P: addr += Offset*U
        if not P or W: REG[Rn] = (REG[Rn] + Offset*U) & 0xFFFFFFFF
        if L: REG[Rd] = mem_read(addr, 4 - 3*B)
        else: mem_write(addr, REG[Rd], 4 - 3*B)
    else:
        Rm = Offset & 15
        # B is Immediate offset flag when D is set
        if B: Offset = (Offset >> 8 & 15) << 4 | Offset & 15
        else: Offset = REG[Rm]
        if P: addr += Offset*U
        if not P or W: REG[Rn] = (REG[Rn] + Offset*U) & 0xFFFFFFFF
        if Typ == 0:
            temp = REG[Rm]  # in case Rm == Rd
            REG[Rd] = mem_read(REG[Rn],2-B)
            mem_write(REG[Rn],temp,2-B)
        elif L:
            if Typ == 1: REG[Rd] = mem_read(addr,2)
            elif Typ == 2: REG[Rd] = mem_read(addr, 1, True)
            elif Typ == 3: REG[Rd] = mem_read(addr, 2, True)
        else:
            if Typ == 1: mem_write(addr,REG[Rd],2)
            elif Typ == 2: REG[Rd],REG[Rd+1] = mem_read(addr,4), mem_read(addr + 4,4)
            elif Typ == 3: mem_write(addr,REG[Rd],4); mem_write(addr + 4,REG[Rd+1],4)


def blocktransfer(instr):
    Flags,Rn,Rlist = instr>>20 & 31, instr>>16 & 15, instr & 0xFFFF
    #Flags: Pre/Post, Up/Down, Load PSR, Writeback, Load/Store
    P,U,S,W,L = (1 if Flags & 2**(4-i) else 0 for i in range(5))
    direction = (-1,1)[U]
    addr = REG[Rn]
    if P: addr += 4*direction
    index = (15,0)[U]
    for i in range(16):
        if Rlist & 1<<index:
            if L: REG[index] = mem_read(addr,4)
            else: mem_write(addr, REG[index],4)
            addr += 4*direction
        index += direction
    if P: addr -= 4*direction
    if W: REG[Rn] = addr


arm_tree = {
    0:(27,36), 1:(26,31), 2:(25,28), 3:(4,9), 4:([25<<20,16<<20],6), 5:dataprocess, 6:(7,8), 7:psr, 
    8:multiply, 9:(7,19), 10:([25<<20,16<<20],12), 11:dataprocess, 12:(6,16), 13:(22,15), 14:arm_bx, 15:clz, 
    16:(5,18), 17:undef, 18:undef, 19:([3<<5,0],23), 20:(22,22), 21:datatransfer, 22:datatransfer, 23:(24,27), 
    24:(23,26), 25:multiply, 26:multiply, 27:datatransfer, 28:([25<<20,16<<20],30), 29:dataprocess, 30:psr, 31:(25,33), 
    32:datatransfer, 33:(4,35), 34:datatransfer, 35:undef, 36:(26,40), 37:(25,39), 38:blocktransfer, 39:arm_branch,
    40:(25,44), 41:([15<<21,2<<21],43), 42:undef, 43:undef, 44:(24,48), 45:(4,47), 46:undef, 47:undef,
    48:undef
}


def navigateTree(instr,tree):
    treepos = 0
    while True:
        try: 
            condition = tree[treepos][0]
            if type(condition) is int:
                if instr>>condition & 1: treepos = tree[treepos][1]
                else: treepos += 1
            else:
                bitmask,value = condition
                if instr & bitmask == value: treepos = tree[treepos][1]
                else: treepos += 1
        except TypeError:
            return tree[treepos]


def execute(instr,mode):
    global BreakState, Executing
    BreakState = ""
    Executing = True
    # THUMB
    if mode:
        REG[15] += 2
        ID = bisect_right(ThumbBounds,instr)
        ThumbFuncs[ID](instr)
    # ARM
    else:
        REG[15] += 4
        Cond = instr >> 28
        if conditions[Cond](REG[16]>>28):
            arm_function = navigateTree(instr,arm_tree)
            arm_function(instr)
    Executing = False